src/nft/components/details/detailsV2/DataPageTraits.tsx

```diff
@@ -1,16 +1,104 @@
 import { Trans } from '@lingui/macro'
-import { GenieAsset } from 'nft/types'
-import { useMemo } from 'react'
+import Column from 'components/Column'
+import { ScrollBarStyles } from 'components/Common'
+import Row from 'components/Row'
+import { GenieAsset, Trait } from 'nft/types'
+import { useMemo, useState } from 'react'
 import styled from 'styled-components/macro'
+import { ThemedText } from 'theme'
+import { opacify } from 'theme/utils'
 
 import { Tab, TabbedComponent } from './TabbedComponent'
+import { TraitRow } from './TraitRow'
 
-const TraitsContentContainer = styled.div`
-  height: 492px;
+const TraitsHeaderContainer = styled(Row)`
+  padding-right: 12px;
 `
 
-const TraitsContent = () => {
-  return <TraitsContentContainer>Traits Content</TraitsContentContainer>
+const TraitsHeader = styled(ThemedText.SubHeaderSmall)<{ $flex?: number; alignRight?: boolean }>`
+  display: flex;
+  line-height: 20px;
+  color: ${({ theme }) => theme.textSecondary};
+  flex: ${({ $flex }) => $flex ?? 1};
+  ${({ alignRight }) => alignRight && 'justify-content: flex-end'};
+`
+
+const TraitRowContainer = styled.div`
+  position: relative;
+`
+
+const TraitRowScrollableContainer = styled.div`
+  overflow-y: auto;
+  overflow-x: hidden;
+  max-height: 412px;
+  width: calc(100% + 6px);
+
+  ${ScrollBarStyles}
+`
+
+const Scrim = styled.div<{ isBottom?: boolean }>`
+  position: absolute;
+  height: 88px !important;
+  left: 0px;
+  right: 12px;
+  ${({ isBottom }) => !isBottom && 'top: 0px'};
+  ${({ isBottom }) => !isBottom && 'transform: matrix(1, 0, 0, -1, 0, 0)'};
+
+  ${({ isBottom }) => isBottom && 'bottom: 0px'};
+
+  background: ${({ theme }) =>
+    `linear-gradient(180deg, ${opacify(0, theme.backgroundSurface)} 0%, ${theme.backgroundSurface} 100%)`}};
+  display: flex;
+`
+
+const TraitsContent = ({ traits }: { traits?: Trait[] }) => {
+  const [userCanScroll, setUserCanScroll] = useState(false)
+  const [scrollProgress, setScrollProgress] = useState(0)
+  const scrollRef = (node: HTMLDivElement) => {
+    if (node !== null) {
+      const canScroll = node.scrollHeight > node.clientHeight
+      canScroll !== userCanScroll && setUserCanScroll(canScroll)
+    }
+  }
+
+  const scrollHandler = (event: React.UIEvent<HTMLDivElement>) => {
+    const scrollTop = event.currentTarget.scrollTop
+    const containerHeight = event.currentTarget.clientHeight
+    const scrollHeight = event.currentTarget.scrollHeight
+
+    setScrollProgress(scrollTop ? ((scrollTop + containerHeight) / scrollHeight) * 100 : 0)
+  }
```

Comment:
I know this same code is used in bag - I think it might be good to make this into its own hook then later we can also refactor the bag to use the same hook

Reply:
great suggestion, I've added the `useSubscribeScrollState` hook